astropy/utils/data.py

```diff
@@ -1311,59 +1411,332 @@ def _get_download_cache_locs():
+        `clear_download_cache` to remove the offending item.
+    OSError, RuntimeError
+        To indicate some problem with the cache structure. This may need a full
+        `clear_download_cache` to resolve, or may indicate some kind of
+        misconfiguration.
+    """
+    # We want full powers over the cache to carry out a proper check
+    with _cache() as (dldir, url2hash):
+        if dldir is None:
+            raise OSError("Cache directory cannot be created or accessed")
+        nonexistent_targets = {}
+        bad_locations = {}
+        bad_hashes = {}
+        abandoned_files = set()
+        leftover_files = set(os.path.join(dldir, k)
+                             for k in os.listdir(dldir))
+        leftover_files.discard(os.path.join(dldir, "urlmap"))
+        leftover_files.discard(os.path.join(dldir, "lock"))
+        for u, h in url2hash.items():
+            if not os.path.exists(h):
+                nonexistent_targets[u] = h
+            leftover_files.discard(h)
+            d, hexdigest = os.path.split(h)
+            if dldir != d:
+                bad_locations[u] = h
+            if check_hashes:
+                hexdigest_file = compute_hash(h)
+                if hexdigest_file != hexdigest:
+                    bad_hashes[u] = h
+        for h in leftover_files:
+            h_base = os.path.basename(h)
+            if len(h_base) >= 32 and re.match(r"[0-9a-f]+", h_base):
+                abandoned_files.add(h)
+        leftover_files -= abandoned_files
+
+        msgs = []
+        if nonexistent_targets:
+            msgs.append(
+                f"URL(s) point(s) to nonexistent file(s): {nonexistent_targets}")
+        if bad_locations:
+            msgs.append(
+                f"URL(s) point(s) to file(s) outside {dldir}: {bad_locations}")
+        if bad_hashes:
+            msgs.append(
+                f"Filename(s) does not match hash(es) of contents: {bad_hashes}")
+        if abandoned_files:
+            msgs.append(
+                f"Apparently abandoned files: {abandoned_files}")
+        if msgs:
+            raise CacheDamaged(
+                "\n".join(msgs),
+                bad_urls=(list(nonexistent_targets.keys())
+                          + list(bad_locations.keys())),
+                bad_files=(list(bad_hashes.values())
+                           + list(abandoned_files)))
         else:
-            return
-    msg = ("Unable to acquire lock for cache directory ({0} exists). "
-           "You may need to delete the lock if the python interpreter wasn't "
-           "shut down properly.")
-    raise RuntimeError(msg.format(lockdir))
+            return leftover_files
 
 
-def _release_download_cache_lock():
-    lockdir = os.path.join(_get_download_cache_locs()[0], 'lock')
+def import_file_to_cache(url_key, filename,
+                         hexdigest=None,
+                         remove_original=False):
+    """Import the on-disk file specified by filename to the cache.
```

Comment:
Minor clarification.
```suggestion
    """Import the on-disk file specified by filename to the cache.
    If the cache already exists with the same content, its timestamp will be updated
    but no copy/move will be done.
```